### PR TITLE
Add IAM support for pubsub topic

### LINKED ISSUE #83: Add IAM support for pubsub topic
_This issue was originally opened by @pdecat as hashicorp/terraform#13602. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi, I would love to see the addition of a `google_pubsub_topic_iam_policy` resource to manage IAM policies on PubSub topics:

Following is an example of what this would look like setting an IAM policy on a pubsub topic:

```
resource "google_service_account" "publisher" {
  account_id   = "publisher"
  display_name = "Publisher"
}

resource "google_pubsub_topic" "default" {
  name = "default-topic"
}

data "google_iam_policy" "topic_publisher" {
  binding {
    role = "roles/pubsub.publisher"

    members = [
      "serviceAccount:${google_service_account.publisher.email}",
    ]
  }
}

# THE FOLLOWING RESOURCE DOES NOT EXIST YET!
resource "google_pubsub_topic_iam_policy" "default_topic_publisher" {
  target      = "${google_pubsub_topic.default.name}"
  policy_data = "${data.google_iam_policy.topic_publisher.policy_data}"
}
```

PS: If this feature request is accepted, should I open other issues for the other resources I am willing to see support? Or should this issue be renamed and updated to track everything in one place?

### PR DESCRIPTION
**Pull request by rosbo**
Fixes #83 

cc @ndmckinley (Can't assign you until you re-accept the contributor invite).

### CODE REVIEW

**Comment by rosbo** [comment]
```sh
--- PASS: TestAccPubsubTopicIamPolicy (4.25s)
--- PASS: TestAccPubsubTopicIamMember (5.49s)
--- PASS: TestAccPubsubTopicIamBinding (7.41s)
```

File: google/resource_pubsub_topic_iam_test.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+	topic := "test-topic-iam-" + acctest.RandString(10)
+	account := "test-topic-iam-" + acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				// Test Iam Member creation (no update for member, no need to test)
```
Comment (danawillow) [inline google/resource_pubsub_topic_iam_test.go]: This comment is misplaced.
Also, should we have an update test for this?

File: google/iam_pubsub_topic.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+func (u *PubsubTopicIamUpdater) GetMutexKey() string {
+	return fmt.Sprintf("iam-folder-%s", u.topic)
+}
+
+func (u *PubsubTopicIamUpdater) DescribeResource() string {
+	return fmt.Sprintf("folder %q", u.topic)
+}
+
+// v1 and v2beta policy are identical
+func resourceManagerToPubsubPolicy(in *cloudresourcemanager.Policy) (*pubsub.Policy, error) {
```
[the remark this answers is not in the document]
Comment (rosbo) [reply google/iam_pubsub_topic.go]: They use the same format:
- resourceManagerToStoragePolicy resourceManagerToPubsubPolicy
- storageToResourceManagerPolicy vs pubsubToResourceManagerPolicy

File: google/iam_pubsub_topic.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+func (u *PubsubTopicIamUpdater) GetResourceId() string {
+	return u.topic
+}
+
+func (u *PubsubTopicIamUpdater) GetMutexKey() string {
+	return fmt.Sprintf("iam-folder-%s", u.topic)
+}
+
+func (u *PubsubTopicIamUpdater) DescribeResource() string {
+	return fmt.Sprintf("folder %q", u.topic)
```
[the remark this answers is not in the document]
Comment (rosbo) [reply google/iam_pubsub_topic.go]: Done

File: google/iam_pubsub_topic.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+	return nil
+}
+
+func (u *PubsubTopicIamUpdater) GetResourceId() string {
+	return u.topic
+}
+
+func (u *PubsubTopicIamUpdater) GetMutexKey() string {
+	return fmt.Sprintf("iam-folder-%s", u.topic)
```
[the remark this answers is not in the document]
Comment (rosbo) [reply google/iam_pubsub_topic.go]: I keep forgetting this one... Done

File: google/resource_pubsub_topic_iam_test.go (hunk 1 above)
Comment (rosbo) [reply google/resource_pubsub_topic_iam_test.go]: Done

File: google/iam_pubsub_topic.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danawillow) [reply google/iam_pubsub_topic.go]: ah sorry I meant in using named return variables:
```
func resourceManagerToStoragePolicy(p *cloudresourcemanager.Policy) (policy *storage.Policy, err error) {
	policy = &storage.Policy{}
	err = Convert(p, policy)
	return
}
```

[the remark this answers is not in the document]
Comment (rosbo) [reply google/iam_pubsub_topic.go]: Done. I kept the new way of doing it and changed iam storage bucket. The error message is more descriptive.

[the remark this answers is not in the document]
Comment (danawillow) [reply google/iam_pubsub_topic.go]: You'll also want to do it for iam_service_account.go and iam_kms_key_ring.go

[the remark this answers is not in the document]
Comment (rosbo) [reply google/iam_pubsub_topic.go]: Done

**Comment by sachams** [comment]
Just been looking for a way to permission pubsub topics! Do you have an idea when this feature might be released?

**Pull request merged**